### PR TITLE
Convert TimePicker to Material 3

### LINKED ISSUE #101480: Update `TimePicker` to support Material 3
https://m3.material.io/components/dialogs/guidelines#754102ae-39d7-47bb-9694-d1ef192beffc

| Preview | 
| --------------- | 
| 	|

Specs: https://m3.material.io/components/time-pickers/overview

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This converts the time picker to use the M3 spec.

Screen Shot 2022-12-01 at 7 35 22 PM

In the process, I had to rearrange a lot of the architecture of the `TimePicker`, since it didn't lend itself to the kinds of layout that the M3 spec wanted.

This PR includes these changes:
* Add a `_TimePickerDefaultsM2` class that contains the defaults for the M2 version of the picker using M3 attributes.
* Add a base `_TimePickerDefaults` class that both `_TimePickerDefaultsM2` and `_TimePickerDefaultsM3` derive from so that they can always return non-nullable values.
* Instead of threading a number of modifiable parameters through all of the widgets, use a single `InheritedModel` that contains the state for the picker.
* Add the M3 defaults class `_TimePickerDefaultsM3` and the appropriate generation template.
* Fixed layout issues where small window sizes caused overflow exceptions.  It resizes nicely now until it reaches a minimum size, and the scrolls the contents after that.
* Added a bunch of material3 specific tests.
* Added a `show_time_picker.0.dart` usage example that allows configuring of the time picker before showing it.
* Reorganized things so that there's a `_TimePicker` widget that could be made public if we want to allow embedding of time pickers into other widgets (instead of in a dialog).

I had to "invent" the token values for the text input fields of the "time input" controls, since while there are values for the hour/minute selector, there's nothing in there for the input field focus colors, borders, etc.

The updated picker handles small windows much better...
<video src="https://user-images.githubusercontent.com/8867023/205209754-8fb43ff5-a65a-4493-ac01-fda1d087ac2a.mp4" controls="controls" style="max-width: 730px;">
</video>

## Related Issues
 - Fixes https://github.com/flutter/flutter/issues/101480

## Tests
 - Added tests for many Material3 attributes and layouts.
 - Added a test for the new example.

### CODE REVIEW

**Comment by gspencergoog** [comment]
@darrenaustin Feel free to start reviewing this: it's only a draft because it's missing a test for the example.

I'm sure you'll find plenty of other things to review here... Sorry for the size.

**Comment by gspencergoog** [comment]
Actually, I think I'll try and factor out and land some of the unrelated changes before I take this off of "draft".

**Comment by gspencergoog** [comment]
Okay, I've factored out and landed the unrelated changes.

File: packages/flutter/lib/src/material/time_picker.dart [hunk 1]
```diff
@@ -2534,3 +3049,664 @@ Future<TimeOfDay?> showTimePicker({
 void _announceToAccessibility(BuildContext context, String message) {
   SemanticsService.announce(message, Directionality.of(context));
 }
+
+// An abstract base class for the M2 and M3 defaults below, so that their return
+// types can be non-nullable.
+abstract class _TimePickerDefaults extends TimePickerThemeData {
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/time_picker.dart:3055]: Interesting. In the other components we just went with a direct subclass of the component theme data. For the few cases where the default had to end in a non-null value, we would just use the '!' operator when pulling it from the defaults. Not as clean perhaps, but reduces the amount of bloat in having yet another full subclass definition in each component.

However, in this case it looks like you are also adding new fields that aren't part of the theme data. Is there some reason those new fields shouldn't be part of the public theme?

File: packages/flutter/lib/src/material/time_picker.dart [hunk 2]
```diff
@@ -1915,168 +2178,491 @@ class TimePickerDialog extends StatefulWidget {
...
+        } else {
+          defaultTheme = _TimePickerDefaultsM2(context);
+          double timePickerWidth;
+          if (alwaysUse24HourFormat) {
+            timePickerWidth = _kTimePickerInputSize.width - defaultTheme.dayPeriodPortraitSize.width - 12;
+          } else {
+            timePickerWidth = _kTimePickerInputSize.width;
+          }
+          timePickerSize = Size(timePickerWidth, _kTimePickerInputSize.height);
+        }
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/time_picker.dart:2369]: This looks duplicated. The only difference between M2 and M3 is just the defaults, right? 

Can't we just use the right default theme (M2/M3) and use the same block of code for both M3 and M2? Or am I missing something that is different in the two cases?

File: packages/flutter/lib/src/material/time_picker.dart [hunk 3]
```diff
@@ -1915,168 +2178,491 @@ class TimePickerDialog extends StatefulWidget {
...
+          );
+        }),
+      ),
+    );
+  }
+}
+
+// The _TimePicker widget is constructed so that in the future we could expose
+// this as a public API for embedding time pickers into other non-dialog
+// widgets, once we're sure we want to support that.
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/time_picker.dart:2527]: Nice. We have in the past seen similar requests for exposing parts of the date picker as well.

File: packages/flutter/lib/src/material/time_picker_theme.dart [hunk 4]
```diff
@@ -133,117 +162,148 @@ class TimePickerThemeData with Diagnosticable {
...
-  ///
-  /// If this is null, the time picker defaults to the overall theme's
-  /// [TextTheme.titleMedium].
-  final TextStyle? dayPeriodTextStyle;
+  /// Deprecated. Use [headlineTextStyle] instead.
+  @Deprecated(
+    'Use headlineTextStyle instead. '
+    'This feature was deprecated after v3.6.0-0.0.pre.',
+  )
+  final TextStyle? helpTextStyle;
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/time_picker_theme.dart:195]: Do we have a flutter fix for this and a migration guide? Or do we want to handle that in a separate PR? Or is this an existing deprecation that just got moved around in this cleanup? 😀

File: packages/flutter/test/material/time_picker_test.dart [hunk 5]
```diff
@@ -358,989 +234,1240 @@ void _tests() {
...
+      await startPicker(tester, (TimeOfDay? time) {
+        expect(find.text('Select time'), findsOneWidget);
+        expect(find.text('Enter time'), findsOneWidget);
+        expect(find.text('Cancel'), findsOneWidget);
+      }, useMaterial3: true);
+    });
 
-    semantics.dispose();
-  });
+    testWidgets('Material2 has ALL CAPS labels', (WidgetTester tester) async {
```
Comment (darrenaustin) [inline packages/flutter/test/material/time_picker_test.dart:246]: For M2 specific tests it would be good to hardwire `useMaterial3: false` so that when we change the default to true they will still work properly.

File: packages/flutter/test/material/time_picker_test.dart [hunk 6]
```diff
@@ -358,989 +234,1240 @@ void _tests() {
...
-    expect(find.text(cancelText), findsOneWidget);
-    expect(find.text(confirmText), findsOneWidget);
-    expect(find.text(helperText), findsOneWidget);
+        await tester.pumpAndSettle();
+        expect(tester.getTopLeft(find.byType(TimePickerDialog)), Offset.zero);
+        expect(tester.getBottomRight(find.byType(TimePickerDialog)), const Offset(390, 600));
+      });
+    });
+
+    group('Works for various view sizes', () {
```
Comment (darrenaustin) [inline packages/flutter/test/material/time_picker_test.dart:773]: Nice!

**Review by darrenaustin — COMMENTED** [review]
Wow, nice work! There is a lot here, and some of the diffs were a little hard to follow, so my apologies if I missed something. Some questions and comments below. Thx!

File: dev/tools/gen_defaults/lib/time_picker_template.dart [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+
+  @override
+  ShapeBorder get shape {
+    return ${shape("$tokenGroup.container")};
+  }
+}
+''';
+
+  /// Generate a [BorderSide] for the given components.
+  String mergedBorder(String componentToken1, String componentToken2) {
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply dev/tools/gen_defaults/lib/time_picker_template.dart:351]: Hmm, I guess not.  Removed.

File: packages/flutter/lib/src/material/time_picker.dart (hunk 3 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/time_picker.dart:2527]: Yeah, I figured I would arrange it this way, and if/when we decide to make it available seprately, we could do the date picker at the same time.

File: packages/flutter/lib/src/material/time_picker.dart (hunk 2 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/time_picker.dart:2369]: Good catch, I didn't notice that!  I consolidated it.

File: packages/flutter/lib/src/material/time_picker.dart [hunk 8]
```diff
@@ -86,186 +79,203 @@ enum TimePickerEntryMode {
...
-                    _DayPeriodControl(
-                      selectedTime: selectedTime,
-                      orientation: orientation,
-                      onChanged: onChanged,
-                    ),
-                  ],
+                ),
+                if (hourDialType == _HourDialType.twelveHour &&
+                    timeOfDayFormat != TimeOfDayFormat.a_space_h_colon_mm) ...<Widget>[
+                  const SizedBox(width: 12),
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/time_picker.dart:246]: Reformatted.

File: packages/flutter/lib/src/material/time_picker.dart (hunk 1 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/time_picker.dart:3055]: I made the separate class with non-null fields so that I could use it everywhere as the type of the resolved default theme, then I wouldn't need to use the `!` everwhere that the default theme is used. It felt a lot cleaner to me, and the intermediate class is pretty much empty.

The sizes seemed like either a bunch of sizes that weren't part of the spec as being customizable, but were part of the token information, or which consolidated the sizing information that was encoded as a bunch of constants in the file that had to be different for M2 and M3, so it seemed better to just use the defaults for those, since they have to be different for M2 and M3 as well.

So, instead of:

```dart
const double _kDayPeriodInputWidth = xx;
const double _kDayPeriodInputHeight = xx;
const double _kDayPeriodInputWidthM2 = xx;
const double _kDayPeriodInputHeightM2 = xx;
```

I could just refer to the sizes in the defaults.

File: packages/flutter/lib/src/material/time_picker.dart [hunk 9]
```diff
@@ -86,186 +79,203 @@ enum TimePickerEntryMode {
...
+
+  static void setSelectedTime(BuildContext context, TimeOfDay value) => of(context, _TimePickerAspect.onSelectedTimeChanged).onSelectedTimeChanged(value);
+  static void setHourMinuteMode(BuildContext context, _HourMinuteMode value) => of(context, _TimePickerAspect.onHourMinuteModeChanged).onHourMinuteModeChanged(value);
 
-  void _handleChangeMode(_TimePickerMode value) {
-    if (value != mode) {
-      onModeChanged(value);
+  @override
+  bool updateShouldNotifyDependent(_TimePickerModel oldWidget, Set<_TimePickerAspect> dependencies) {
+    if (use24HourFormat != oldWidget.use24HourFormat && dependencies.contains(_TimePickerAspect.use24HourFormat)) {
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/time_picker.dart:151]: Oh, absolutely.

File: packages/flutter/test/material/time_picker_test.dart (hunk 5 above)
Comment (gspencergoog) [reply packages/flutter/test/material/time_picker_test.dart:246]: Oh, good point. Done.

File: packages/flutter/lib/src/material/time_picker_theme.dart (hunk 4 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/time_picker_theme.dart:195]: No, it's a new deprecation, and I think I'd rather handle the flutter fix in another PR.
Or, I could just duplicate the value here, and deprecate it in the PR that has the fix.

Comment (gspencergoog) [reply packages/flutter/lib/src/material/time_picker_theme.dart:195]: Actually, I'll just not deprecate it at all in this PR.  We can do that later.

Comment (darrenaustin) [reply packages/flutter/lib/src/material/time_picker_theme.dart:195]: Either would be fine. I just wanted to make sure we have a plan to follow through on the deprecation. Thx.

**Comment by gspencergoog** [comment]
Thanks for that huge review!

Okay, I think I addressed all the comments, PTAL. For the test, I actually rearranged it so that a) it's explicit whether things are set up as material2 or material3, and I put a loop around the test definitions so that we actually run every test once in material2 mode and once in material3 mode.

File: packages/flutter/lib/src/material/time_picker.dart (hunk 1 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/time_picker.dart:3040]: That makes sense. Early on in the migration we got some push back on adding more classes to each widget as it was inflating the download size for the web. I believe each dart class gets transpiled into a JS class, so that could be significant for these boilerplate heavy classes. That said in this case it probably doesn't matter that much and ti is cleaner. Thx.

**Review by darrenaustin — APPROVED** [review]
flutter-lgtm

Nice work. Thx.
